source/common/stats/thread_local_store.cc

```diff
@@ -233,6 +235,10 @@ Counter& ThreadLocalStoreImpl::ScopeImpl::counter(const std::string& name) {
   // Determine the final name based on the prefix and the passed name.
   std::string final_name = prefix_ + name;
 
+  if (parent_.stats_matcher_->rejects(final_name)) {
```

Comment:
The only downside I see to this approach is that we will need to consult this rejection list every time we do a dynamic stat lookup for things like HTTP response code stats. I could see this potentially becoming a performance bottleneck in certain configs, or maybe it doesn't matter. Did you give any thought to this case out of curiosity?

Reply:
Very good point. I think there's not an issue with a positive result -- that will be cached in ThreadLocalStore. However there's an issue with negative results: if we are rejecting all stats with a regex, we'll have to re-reject them many times per request e.g. here: 

https://github.com/envoyproxy/envoy/blob/7ccf67db45db8c4e9bc406488c8f1f2a8566f506/source/common/http/codes.cc#L21

One possible resolution is to cache negative results in thread_local_store.cc as well. It's only important in the data-path, but as of now you don't know if the lookup is data-path-critical. @mattklein123 suggested in https://github.com/envoyproxy/envoy/issues/4196#issuecomment-425670941 to add attributes to the stats indicating whether they would be looked up in the data path, and cache rejections for those only.

Depending on what stats-matching has been requested it might not be worth it: e.g. a matcher in "accept all" or "reject all" mode would probably be OK, and maybe even "reject/accept prefix" if there's just ~1 substring. But regexes would be a drag.

Reply:
Actually this will filter before the cache. Maybe we should put the matcher after the cache lookup?

Reply:
I don't think you can move it from where it is, otherwise the stat will be in the central store and then start showing up during flushing, admin output, etc.

TBH I think this is fine for now and we can iterate (yes there will be some CPU burned checking for an empty rejection list in the hot path but I would probably take that in the interest of moving this forward). However, can you potentially add some comments/warnings above the three `rejects` calls that we might need to think about this more in the future if perf becomes an issue (optimally just put the comment in one place and refer the others to the single comment)?

Reply:
Agreed; maybe also a warning in doc that the regex form of stat matcher may impact perf.

Reply:
Thanks all, I've added more warnings in c9f8d25.